airflow/jobs/scheduler_job_runner.py

```diff
@@ -1844,6 +1895,58 @@ def _executor_to_tis(self, tis: list[TaskInstance]) -> dict[BaseExecutor, list[T
                 executor = str(ti.executor)
             else:
                 executor = None
-            _executor_to_tis[ExecutorLoader.load_executor(executor)].append(ti)
+
+            if executor_obj := self._try_to_load_executor(executor):
+                _executor_to_tis[executor_obj].append(ti)
+            else:
+                continue
 
         return _executor_to_tis
+
+    def _slots_free_for_tis(self, tis: list[TaskInstance]) -> set[TaskInstance]:
+        """Return TIs that we have slots available for."""
+        # First get a mapping of executor names to slots they have available
+        executor_to_slots_available: dict[ExecutorName, int] = {}
+        for executor in self.job.executors:
+            # All executors should have a name if they are initted from the executor_loader. But we need to
+            # check for None to make mypy happy.
+            if executor.name:
+                executor_to_slots_available[executor.name] = executor.slots_available
+            else:
+                raise AirflowException(f"Executor {executor} did not have a `name` field configured!")
+
+        # Loop through all the TIs we're scheduling for and add them to a set to be moved to queued if the
+        # executor they're assigned to has slots available.
+        tis_we_have_room_for = set()
+        for ti in tis:
+            if ti.executor:
+                executor = str(ti.executor)
+            else:
+                executor = None
+            if executor_obj := self._try_to_load_executor(executor):
+                ti_exec_name = executor_obj.name
+            else:
+                continue
```

Comment:
shouldn't you fail the task in this case?

Reply:
i know it's cosmetic but it just seems like some of the ifs are probably unnecessary

seems like the logic should be something like 

```python
    for ti in tis:
        try:
            executor = self._try_to_load_executor(ti.executor)
        except CouldNotLoadExecutor:
            fail_task(ti)
            continue    
        if executor_to_slots_available[executor.name] > 0:
            tis_we_have_room_for.add(ti)
            executor_to_slots_available[executor.name] -= 1
```

Reply:
> shouldn't you fail the task in this case?

If we arrive at this case the _try_to_load_executor call would have already logged to the user that the task can't be run and they need to reconfigure it or their executor (there is also a UI dialogue thrown now as well). So we just continue on to the next TI. This is the behaviour agreed in the AIP.

> seems like the logic should be something like

We need the exec name not the executor since that's what the TIs have and is needed in the scheduling logic where this helper is called, it's easier to use as a dict key too if you're debugging/logging.
Also we don't want to fail as mentioned above.
I think this code just looks a bit intimidating because of the mypy stuff. Otherwise it's basically as simplified as you have it. I'd like to keep it as is if you don't mind! :pray:

Reply:
> I'd like to keep it as is if you don't mind! 🙏

as usual, just a suggestion :)